homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,324 @@
+        entity (string): used for debugging only
+    """
+
+    def __init__(self, name, window_size=1, precision=None, entity=None):
+        """Initialize common attributes."""
+        self.states = deque(maxlen=window_size)
+        self.precision = precision
+        self._stats = {}
+        self._name = name
+        self._entity = entity
+        self._skip = False
+
+    @property
+    def name(self):
+        """Return filter name."""
+        return self._name
+
+    @property
+    def skip(self):
+        """Return wether the current filter_state should be skipped."""
+        return self._skip
+
+    @property
+    def stats(self):
+        """Return statistics of the filter."""
+        return self._stats
+
+    def _filter_state(self, new_state):
+        """Implement filter."""
+        raise NotImplementedError()
+
+    def filter_state(self, new_state):
+        """Implement a common interface for filters."""
+        filtered = self._filter_state(new_state)
+        if self.precision is None:
+            return filtered
+        return round(filtered, self.precision)
+
+
+class OutlierFilter(Filter):
+    """BASIC outlier filter.
+
+    Determines if new state is in a band around the median.
+
+    Args:
+        radius (float): band radius
+    """
+
+    def __init__(self, window_size, precision, entity, radius):
+        """Initialize Filter."""
+        super().__init__(FILTER_NAME_OUTLIER, window_size, precision, entity)
+        self._radius = radius
+        self._stats_internal = Counter()
+
+    def _filter_state(self, new_state):
+        """Implement the outlier filter."""
+        self._stats_internal['total_filtered'] += 1
+        new_state = float(new_state)
+
+        self._stats['erasures'] = "{0:.2f}%".format(
+            100 * self._stats_internal['erasures']
+            / self._stats_internal['total_filtered']
+        )
+
+        if (len(self.states) > 1 and
+                abs(new_state - statistics.median(self.states))
+                > self._radius):
+
+            self._stats_internal['erasures'] += 1
+
+            _LOGGER.debug("Outlier in %s: %s", self._entity, new_state)
+            return self.states[-1]
+        return new_state
+
+
+class LowPassFilter(Filter):
+    """BASIC Low Pass Filter.
+
+    Args:
+        time_constant (int): time constant.
+    """
+
+    def __init__(self, window_size, precision, entity, time_constant):
+        """Initialize Filter."""
+        super().__init__(FILTER_NAME_LOWPASS, window_size, precision, entity)
+        self._time_constant = time_constant
+
+    def _filter_state(self, new_state):
+        """Implement the low pass filter."""
+        new_state = float(new_state)
+
+        try:
+            new_weight = 1.0 / self._time_constant
+            prev_weight = 1.0 - new_weight
+            filtered = prev_weight * self.states[-1] + new_weight * new_state
+        except IndexError:
```

Comment:
Prefer to just catch an empty states list instead of going for `try…except`.

```python
if not self.states
    return new_state

new_weight = …
```